pkg/distributor/distributor_test.go

```diff
@@ -2339,6 +2345,9 @@ func TestDistributor_IngestionIsControlledByForwarder(t *testing.T) {
 			ingestSample:          false,
 			expectIngestedMetrics: []string{},
 			expectedMetrics: `
+			# HELP cortex_distributor_received_requests_total The total number of received requests, excluding rejected, forwarded and deduped requests.
+			# TYPE cortex_distributor_received_requests_total counter
+			cortex_distributor_received_requests_total{user="user"} 1
```

Comment:
I was surprised by this result: at first I thought that there shouldn't actually be any request making it through to the `pushWithCleanup` method if fowarding rules had been applied. However, it seems that we don't short-circuit the request even if all the samples are forwarded and the `req.Timeseries` ends up empty. Thus, I've set the expected values here to match the current behavior. Maybe @replay could shed some light on why this?

Reply:
I'm fine writing the assertion in this PR to match the actual behaviour, but from your description it smells like a bug to me.

Reply:
The reason is because there might still be metadata in the request which we want to ingest, the forwarding never modifies any metadata. If the metadata property is empty then yes we could skip it.

Reply:
I've attempted to short-circuit in case both `req.Timeseries` and `req.Metadata` are empty but the resulting code is a bit more complicated than the current code and the metrics emitted in tests got even more surprising. I'm going to leave the current behavior as-is since it is not hurting anything and we can clean up the behavior of the middlewares and metrics in #2783.